Always Show Waypoints Option / new customization Options / new Waypoi…

slizza-io opened the pull request:
### Changes
•	Small changes in how waypoints are created, loaded, and deleted.
•	New style of waypoints.
•	New class WaypointObjectMeta to store and handle data of the waypoint world object.
•	New listeners GameTickListener to handle how waypoints are rendered and loaded.
•	New widgets (HeaderWidget, WaypointListItemWidget).
•	New utility classes.
•	LabyGradleVersion updated to 0.3.28

### Features
•	Option to show waypoints on the whole map from any position (default: disabled).
•	Option to render waypoints with a background (default: disabled).
•	Option to show marker icons with waypoints (default: enabled).
•	Option to hide individual waypoints.

### Todos (listed by priority)
•	Create a config converter to convert old config to the new one, since the DefaultWaypoint class has been changed.
•	Render Waypoints world-depended -> a unique identifier for the world, where the waypoint got created will be needed, which will be saved in the config, instead of „PLACEHOLDER“ for each waypoint
•	Add more customization options.

slizza-io commented:
### Additional Info

My changes are not production ready until a config converter is written to convert the existing waypoints in the config to the new format. The new format requires the following properties for the already existing waypoints:

- `world`: Set to `"LEGACY_WAYPOINT"`
- `visible`: Set to `true`

derrop commented on api/src/main/java/net/labymod/addons/waypoints/waypoint/Waypoint.java:
```diff
@@ -19,4 +19,9 @@ default Color color() {
+
+  @Override
+  default boolean isSeeThrough() {
+    return true;
```
When enabling this, you should consider handling the "darker" parameter in the renderInWorld method (see DefaultWaypoint). This is to get a similar behavior for player name tags that are rendered through walls but slightly darker.

You can have a look at the MarkerObject class in LabyMod which basically just uses the following color for rendering (alpha 75 when darker is true):
```java
ColorUtil.toValue(255, 255, 255, darker ? 75 : 255)
```

derrop reviewed (changes requested):
First of all thanks for your contribution.
In general, please always use `this.` in front when accessing non-static fields. Also please reformat your code (in IntelliJ you can select the whole project folder on the left handside and press Ctrl Alt L), for example sometimes you are using 2 spaces and sometimes 4 spaces for indentation.

For the config conversion, everything should already be available in LabyMod. You can set the config version in the configuration to 2 and handle the ConfigurationVersionUpdateEvent.

Regarding the world identifier I don't think there is a LabyMod feature yet to access the name of a world so I will be adding something like that soon. In multiplayer you can already access everything via `Laby.references().serverController().getCurrentServerData()`.

slizza-io commented:
### New Changes
- Renamed "GameTickListener" to "WaypointsUpdateListener"
- Improved Waypoint rendering performance by caching Waypoint data
- Fixed glitching Background
